// zod-openapi/lib-es2015/create/schema/tuple.js
import { satisfiesVersion } from '../../openapi';
import { createSchemaOrRef } from '.';
export const createTupleSchema = (zodTuple, state) => {
    const items = zodTuple.items;
    const rest = zodTuple._def.rest;
    return {
        type: 'array',
        ...mapItemProperties(items, rest, state),
    };
};
const mapPrefixItems = (items, state) => {
    if (items.length) {
        return items.map((item) => createSchemaOrRef(item, state));
    }
    return undefined;
};
const mapItemProperties = (items, rest, state) => {
    const prefixItems = mapPrefixItems(items, state);
    if (satisfiesVersion(state.components.openapi, '3.1.0')) {
        if (!rest) {
            return {
                maxItems: items.length,
                minItems: items.length,
                ...(prefixItems && { prefixItems }),
            };
        }
        return {
            items: createSchemaOrRef(rest, state),
            ...(prefixItems && { prefixItems }),
        };
    }
    if (!rest) {
        return {
            maxItems: items.length,
            minItems: items.length,
            ...(prefixItems && { items: { oneOf: prefixItems } }),
        };
    }
    return {
        ...(prefixItems && {
            items: {
                oneOf: [...prefixItems, createSchemaOrRef(rest, state)],
            },
        }),
    };
};
