// zod-openapi/lib-commonjs/create/document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDocumentYaml = exports.createDocumentJson = exports.createDocument = void 0;
const yaml_1 = require("yaml");
const components_1 = require("./components");
const paths_1 = require("./paths");
const createDocument = (zodOpenApiObject) => {
    const components = zodOpenApiObject.components ?? {};
    const defaultComponents = (0, components_1.getDefaultComponents)(components, zodOpenApiObject.openapi);
    return {
        ...zodOpenApiObject,
        paths: (0, paths_1.createPaths)(zodOpenApiObject.paths, defaultComponents),
        webhooks: (0, paths_1.createPaths)(zodOpenApiObject.webhooks, defaultComponents),
        components: (0, components_1.createComponents)(components, defaultComponents),
    };
};
exports.createDocument = createDocument;
const createDocumentJson = (params, jsonOptions) => {
    const document = (0, exports.createDocument)(params);
    return JSON.stringify(document, jsonOptions?.replacer, jsonOptions?.options ?? 2);
};
exports.createDocumentJson = createDocumentJson;
const createDocumentYaml = (params, yamlOptions = {}) => {
    const document = (0, exports.createDocument)(params);
    return (0, yaml_1.stringify)(document, yamlOptions.replacer, yamlOptions.options);
};
exports.createDocumentYaml = createDocumentYaml;
