// zod-openapi/lib-commonjs/create/parameters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createParametersObject = exports.createParamOrRef = exports.createBaseParameter = exports.createComponentParamRef = void 0;
const schema_1 = require("./schema");
const createComponentParamRef = (ref) => `#/components/parameters/${ref}`;
exports.createComponentParamRef = createComponentParamRef;
const createBaseParameter = (schema, components) => {
    const { ref, ...rest } = schema._def.openapi?.param ?? {};
    const schemaOrRef = (0, schema_1.createSchemaOrRef)(schema, {
        components,
        type: 'input',
    });
    const required = !schema.isOptional();
    return {
        ...rest,
        ...(schema && { schema: schemaOrRef }),
        ...(required && { required }),
    };
};
exports.createBaseParameter = createBaseParameter;
const createParamOrRef = (zodSchema, type, name, components) => {
    const component = components.parameters.get(zodSchema);
    if (component && component.type === 'complete') {
        if (!('$ref' in component.paramObject) &&
            (component.paramObject.in !== type || component.paramObject.name !== name)) {
            throw new Error(`parameterRef "${component.ref}" is already registered`);
        }
        return {
            $ref: (0, exports.createComponentParamRef)(component.ref),
        };
    }
    // Optional Objects can return a reference object
    const baseParamOrRef = (0, exports.createBaseParameter)(zodSchema, components);
    if ('$ref' in baseParamOrRef) {
        throw new Error('Unexpected Error: received a reference object');
    }
    const ref = zodSchema?._def?.openapi?.param?.ref ?? component?.ref;
    const paramObject = {
        in: type,
        name,
        ...baseParamOrRef,
    };
    if (ref) {
        components.parameters.set(zodSchema, {
            type: 'complete',
            paramObject,
            ref,
        });
        return {
            $ref: (0, exports.createComponentParamRef)(ref),
        };
    }
    return paramObject;
};
exports.createParamOrRef = createParamOrRef;
const createParameters = (type, zodObject, components) => {
    if (!zodObject) {
        return [];
    }
    return Object.entries(zodObject.shape).map(([key, zodSchema]) => (0, exports.createParamOrRef)(zodSchema, type, key, components));
};
const createRequestParams = (requestParams, components) => {
    if (!requestParams) {
        return [];
    }
    const pathParams = createParameters('path', requestParams.path, components);
    const queryParams = createParameters('query', requestParams.query, components);
    const cookieParams = createParameters('cookie', requestParams.cookie, components);
    const headerParams = createParameters('header', requestParams.header, components);
    return [...pathParams, ...queryParams, ...cookieParams, ...headerParams];
};
const createParametersObject = (parameters, requestParams, components) => {
    const createdParams = createRequestParams(requestParams, components);
    const combinedParameters = [
        ...(parameters ? parameters : []),
        ...createdParams,
    ];
    return combinedParameters.length ? combinedParameters : undefined;
};
exports.createParametersObject = createParametersObject;
