// zod-openapi/lib-commonjs/create/paths.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPaths = exports.createRequestBody = void 0;
const components_1 = require("./components");
const content_1 = require("./content");
const parameters_1 = require("./parameters");
const responses_1 = require("./responses");
const specificationExtension_1 = require("./specificationExtension");
const createRequestBody = (requestBodyObject, components) => {
    if (!requestBodyObject) {
        return undefined;
    }
    const component = components.requestBodies.get(requestBodyObject);
    if (component && component.type === 'complete') {
        return {
            $ref: (0, components_1.createComponentRequestBodyRef)(component.ref),
        };
    }
    const ref = requestBodyObject.ref ?? component?.ref;
    const requestBody = {
        ...requestBodyObject,
        content: (0, content_1.createContent)(requestBodyObject.content, components, 'input'),
    };
    if (ref) {
        components.requestBodies.set(requestBodyObject, {
            type: 'complete',
            ref,
            requestBodyObject: requestBody,
        });
        return {
            $ref: (0, components_1.createComponentRequestBodyRef)(ref),
        };
    }
    return requestBody;
};
exports.createRequestBody = createRequestBody;
const createOperation = (operationObject, components) => {
    if (!operationObject) {
        return undefined;
    }
    const { parameters, requestParams, requestBody, responses, ...rest } = operationObject;
    const maybeParameters = (0, parameters_1.createParametersObject)(parameters, requestParams, components);
    const maybeRequestBody = (0, exports.createRequestBody)(operationObject.requestBody, components);
    const maybeResponses = (0, responses_1.createResponses)(operationObject.responses, components);
    return {
        ...rest,
        ...(maybeParameters && { parameters: maybeParameters }),
        ...(maybeRequestBody && { requestBody: maybeRequestBody }),
        ...(maybeResponses && { responses: maybeResponses }),
    };
};
const createPathItem = (pathObject, components) => {
    const { get, put, post, delete: del, options, head, patch, trace, ...rest } = pathObject;
    const maybeGet = createOperation(get, components);
    const maybePut = createOperation(put, components);
    const maybePost = createOperation(post, components);
    const maybeDelete = createOperation(del, components);
    const maybeOptions = createOperation(options, components);
    const maybeHead = createOperation(head, components);
    const maybePatch = createOperation(patch, components);
    const maybeTrace = createOperation(trace, components);
    return {
        ...rest,
        ...(get && { get: maybeGet }),
        ...(put && { put: maybePut }),
        ...(post && { post: maybePost }),
        ...(del && { delete: maybeDelete }),
        ...(options && { options: maybeOptions }),
        ...(head && { head: maybeHead }),
        ...(patch && { patch: maybePatch }),
        ...(trace && { trace: maybeTrace }),
    };
};
const createPaths = (pathsObject, components) => {
    if (!pathsObject) {
        return undefined;
    }
    return Object.entries(pathsObject).reduce((acc, [path, pathItemObject]) => {
        if ((0, specificationExtension_1.isISpecificationExtension)(path)) {
            acc[path] = pathItemObject;
            return acc;
        }
        acc[path] = createPathItem(pathItemObject, components);
        return acc;
    }, {});
};
exports.createPaths = createPaths;
