// zod-openapi/lib-commonjs/create/responses.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createResponses = exports.createResponse = exports.createComponentHeaderRef = exports.createBaseHeader = exports.createHeaderOrRef = exports.createResponseHeaders = void 0;
const components_1 = require("./components");
const content_1 = require("./content");
const schema_1 = require("./schema");
const specificationExtension_1 = require("./specificationExtension");
const createResponseHeaders = (responseHeaders, components) => {
    if (!responseHeaders) {
        return undefined;
    }
    return Object.entries(responseHeaders.shape).reduce((acc, [key, zodSchema]) => {
        acc[key] = (0, exports.createHeaderOrRef)(zodSchema, components);
        return acc;
    }, {});
};
exports.createResponseHeaders = createResponseHeaders;
const createHeaderOrRef = (schema, components) => {
    const component = components.headers.get(schema);
    if (component && component.type === 'complete') {
        return {
            $ref: (0, exports.createComponentHeaderRef)(component.ref),
        };
    }
    // Optional Objects can return a reference object
    const baseHeader = (0, exports.createBaseHeader)(schema, components);
    if ('$ref' in baseHeader) {
        throw new Error('Unexpected Error: received a reference object');
    }
    const ref = schema._def?.openapi?.header?.ref ?? component?.ref;
    if (ref) {
        components.headers.set(schema, {
            type: 'complete',
            headerObject: baseHeader,
            ref,
        });
        return {
            $ref: (0, exports.createComponentHeaderRef)(ref),
        };
    }
    return baseHeader;
};
exports.createHeaderOrRef = createHeaderOrRef;
const createBaseHeader = (schema, components) => {
    const { ref, ...rest } = schema._def.openapi?.header ?? {};
    const schemaOrRef = (0, schema_1.createSchemaOrRef)(schema, {
        components,
        type: 'input',
    });
    const required = !schema.isOptional();
    return {
        ...rest,
        ...(schema && { schema: schemaOrRef }),
        ...(required && { required }),
    };
};
exports.createBaseHeader = createBaseHeader;
const createComponentHeaderRef = (ref) => `#/components/headers/${ref}`;
exports.createComponentHeaderRef = createComponentHeaderRef;
const createHeaders = (headers, responseHeaders, components) => {
    if (!responseHeaders && !headers) {
        return undefined;
    }
    const createdHeaders = (0, exports.createResponseHeaders)(responseHeaders, components);
    return {
        ...headers,
        ...createdHeaders,
    };
};
const createResponse = (responseObject, components) => {
    if ('$ref' in responseObject) {
        return responseObject;
    }
    const component = components.responses.get(responseObject);
    if (component && component.type === 'complete') {
        return { $ref: (0, components_1.createComponentResponseRef)(component.ref) };
    }
    const { content, headers, responseHeaders, ref, ...rest } = responseObject;
    const maybeHeaders = createHeaders(headers, responseHeaders, components);
    const response = {
        ...rest,
        ...(maybeHeaders && { headers: maybeHeaders }),
        ...(content && { content: (0, content_1.createContent)(content, components, 'output') }),
    };
    const responseRef = ref ?? component?.ref;
    if (responseRef) {
        components.responses.set(responseObject, {
            responseObject: response,
            ref: responseRef,
            type: 'complete',
        });
        return {
            $ref: (0, components_1.createComponentResponseRef)(responseRef),
        };
    }
    return response;
};
exports.createResponse = createResponse;
const createResponses = (responsesObject, components) => Object.entries(responsesObject).reduce((acc, [path, responseObject]) => {
    if ((0, specificationExtension_1.isISpecificationExtension)(path)) {
        acc[path] = responseObject;
        return acc;
    }
    acc[path] = (0, exports.createResponse)(responseObject, components);
    return acc;
}, {});
exports.createResponses = createResponses;
