// zod-openapi/lib-es2015/create/document.js
import { stringify } from 'yaml';
import { createComponents, getDefaultComponents } from './components';
import { createPaths } from './paths';
export const createDocument = (zodOpenApiObject) => {
    const components = zodOpenApiObject.components ?? {};
    const defaultComponents = getDefaultComponents(components, zodOpenApiObject.openapi);
    return {
        ...zodOpenApiObject,
        paths: createPaths(zodOpenApiObject.paths, defaultComponents),
        webhooks: createPaths(zodOpenApiObject.webhooks, defaultComponents),
        components: createComponents(components, defaultComponents),
    };
};
export const createDocumentJson = (params, jsonOptions) => {
    const document = createDocument(params);
    return JSON.stringify(document, jsonOptions?.replacer, jsonOptions?.options ?? 2);
};
export const createDocumentYaml = (params, yamlOptions = {}) => {
    const document = createDocument(params);
    return stringify(document, yamlOptions.replacer, yamlOptions.options);
};
